### ISSUE TITLE
RGP Lua experimental 0.67b3 problem

### ISSUE THREAD

**Issue by cv-on-hub** [finale-lua/lua-scripts#495]
Found a tiny (though fatal) error with RGP Lua 0.67b3 at library/mixin.lua:544 - “attempt to index a boolean value (local ‘meta’)”. Happens on the first call to mixin.FCXCustomLuaWindow() without anything else involved. (affected scripts work fine with 0.66 and 0.67b1)

**Comment by rpatters1** [comment]
I have investigated it, and this is a LuaBridge compatibility issue. By default, it locks down the class metatables. I'll re-issue the LuaBridge3 version later today (hopefully) with that restriction lifted. I'm hoping to undersand the security implications first before I do, and I have a question out to that end.

**Comment by rpatters1** [comment]
@ThistleSifter is there a way to do mixins without access to the metatables? I'm reading that a lot of other frameworks are treating this as a security hole and closing it down. That includes the original LuaBridge in an upcoming version.

**Comment by ThistleSifter** [comment]
Yes it can work, but it would require significant changes.

Doing it without metatables would mean transforming the mixin library from modifications of the original object to wrapper tables. As it stands, mixin-enabled Finale objects can be seamlessly passed to other standard Finale objects (or libraries built around standard Finale objects) without needing to do anything. All type checking passes any tests, since it is the original object being used. Using wrapper tables would make it an "either you're in or you're out" affair, when using mixins.

**Comment by rpatters1** [comment]
That sounds like something we should strive to keep. However, do we know if any of the client code is using it that way?

**Comment by ThistleSifter** [comment]
I don't think so, but that would most likely be because we don't really have any mixins for notation classes yet, they're mostly dialog-related.

**Comment by rpatters1** [comment]
I have uploaded 0.67b4 to the download site. It re-enables access to the metatables. This will allow other testing (like the numerical display issues we were seeing). However, I think we need to assume that access to the metatables will go away at some point. I can't see it being too much trouble for a programmer to have to use an `.fc` property, or some such, to get the original class for sending to the PDK. I just don't know how big a rewrite it is.

**Comment by ThistleSifter** [comment]
I have a fix in the works for the number issue, it should be done in the next day or two.

For the metatables issue, it's mainly planning it out and making sure edge cases are covered. Eg what happens when a non-mixin library function is called that returns another finale object? How can we ensure automatic mixin promotion on the returned values? Etc, etc

**Comment by cv-on-hub** [comment]
Confirming that RGP 0.67b4 does not throw an error on calls to `mixin.FCXCustomLuaWindow()`, but decimal points persist on `dialog.CreateMeasurementEdit()` integers.

**Comment by rpatters1** [comment]
I don't know how much interest you have in how the sausage gets made, but coincidentally [this pull request](https://github.com/kunitoki/LuaBridge3/pull/116) over at LuaBridge3 is going to allow direct class extensibility by Lua code. That is, if you want to add a Lua method to a finale class, then you would just add it, like:

```lua
function finale.FCControl:SetTextString(str)
   local fcstr = finale.FCString()
   fcstr.LuaString = str
   self:SetText(fcstr)
end
```

What it won't do is let you replace existing functions. I realize this doesn't fully replace the flexibility of rewriting the class metatable, but I wonder if it would help in creating a revised `mixin` library that does not need to replace the `__index` and `__newindex` functions.

Anyway, I thought at least @ThistleSifter might be interested. In order to use it, we'd have to commit to adopting LuaBridge3 (which I'm not quire ready to do) and then I'd have to enable extensibility on the finale classes, which could be done easily enough.

**Comment by cv-on-hub** [comment]
I don't understand all of it but am fascinated by the processes, and little bits keep sinking in.

**Comment by rpatters1** [comment]
As the situation is developing, it looks like it will be possible *selectively* to permit access to the metatables. So the question is, how do we identify trusted code (e.g., mixins) that we will allow to modify them vs. untrusted code. A user-specified config option is one possibility. Anyone have other ideas?

**Comment by ThistleSifter** [comment]
Thanks for linking that PR, I just had a quick read of it and it looks promising. I'll leave some comments over there too, but here are some quick thoughts as it relates to what we have here.

For the past couple of days I've been playing with rewriting the mixin library to use wrapping tables, so I've been pondering pros and cons of both approaches.

- Replacing methods at a global level is very convenient but it does mean that we'd have to be more careful about not introducing breaking changes since it would be easy to have unintended consequences (eg throwing an error instead of returning boolean for success/failure).
- Modified methods can also come from multiple sources (bearing in mind personal scripts). I'm not sure how they'd all play nicely together.
- The way the modified methods are loaded would need to be reworked.
- I don't think the fluid interface can be done without access to `__index`.
- FCX mixins would only be able to be implemented if there was a way of subclassing in Lua. I'll ask about that.
- In some ways I like the clear delineation between extension and original object that wrapping tables provide. It allows more freedom in extending without needing to worry about unintended consequences.

As for identifying trusted code...I'm not sure at the moment, I'll have a think.

**Comment by ThistleSifter** [comment]
I'm not sure how a user option would work since it would be the script author that decides if they need the extensions or not. Unless you have a permission dialog that pops up the first time the user runs a script that uses extensions?

**Comment by rpatters1** [comment]
The user option would be an option in the RGP Lua config dialog. Not sure what it could say though.

I've already implemented hash checking for ExecuteAtStartup scripts. Maybe there could be a hash stored on GitHub (or on the finalelua site) that RGP Lua could cross check.

**Comment by rpatters1** [comment]
Following on the last point, we previously discussed the possibility of using hash checking to discover script updates. It could alos be used to authenticate scripts from the finalelua.com site and greenlight the use of `getmetatable` among other things.

**Comment by rpatters1** [comment]
All right, so this unix command generates the same hash that my code is generating internally:

```bash
shasum -a 512 <filename>
```

This outputs the hash followed by the filename. I could strip the filename off and verify that against the filename I've got as well as calculating the SHA-512 hash on my end.

Here is a possible approach. I am open to suggestions for improving it. We should be able to do a couple of things in the deploy script that could make this work, both for identifying trusted code and for detecting when there are updates:

- add a `HashURL` finaleplugin option to the `dist` scripts. (Or maybe just a `CheckTrusted` boolean and let RGP Lua construct the url.) RGP Lua could check this once when it starts up and then just check the hash locally thereafter, so we don't take the performance hit each time we invoke it.
- modify `deploy.yml` to run the above command on every file in the `dist` directory and store that in a separate directory under the name `<filename>.hash` (for example). This would be the url for the HashURL option.
- The HashURL should just be the trailing part of the URL. The front part should be only a trusted repository. As of now, only `https://github.com/finale-lua/lua-scripts`.
- Add a user option `Trusted` to RGP Lua that overrides this and allows any script to be trusted. (For debugging or for third parties that are writing their own code.) In the next release of RGP Lua, the config file is protected with hash-checking and encryption. This presents at least something of a barrier to external code rewriting it. Given that the risk is low, it's probably enough.

I have looked at `deploy.yml`, but I don't really understand it. One thing that confuses me is the `./lua` and `./website` directories, because those directories don't exist in the repo. But it seems like adding the hash-generation part should be easy.

Modifying the `plugindef` functions would go in the other repo, I'm guessing.

**Comment by ThistleSifter** [comment]
The `.lua` and `.website` directories are checkout directories. The deploy action is just a series of commands run on an ubuntu instance. The first thing it does is checkout a copy of this repo to `.lua` and the docs repo to `.website`. Then it does whatever it needs to on its local copy, commits the changes and pushes them back to the central repos. I'm happy to make changes to the deploy action if you need.

How would different script versions be handled? Would we need to maintain a store of the complete hash history for each script?

**Comment by rpatters1** [comment]
I'm thinking we would not need to maintain a hash history. For versioning, the way it could work is that some entity (a script, an external program, or RGP Lua itself) would calculate the hash on the local copy. If it differed from the current value on GitHub, we could infer that there had been a version change.

I'm not sure I want RGP Lua itself to be in charge of version checking, but we can discuss that as needed. I could make the calculated hash for the file available through the `FCLuaScriptItem` interface, and the version checking could be done there.

One downside to this approach is that a local script loses its trusted status if it is updates on GitHub. I could add a hidden option to the config file that marks the hash as verified. Then it would not need to be checked against GitHub again until the contents changed.

**Comment by rpatters1** [comment]
Is the `./website` checked back into a repo on GitHub? I am curious to look at the output.

**Comment by ThistleSifter** [comment]
Yes, that goes here: https://github.com/finale-lua/jw-lua-scripts-docs

**Comment by ThistleSifter** [comment]
So, which direction are you leaning towards for the future of the library (wrapping tables, extensions, metatable, combination of extensions/metatable)?

**Comment by rpatters1** [comment]
I'm thinking if we can get the trusted code thing working, we keep it simple and use the metatables.

**Comment by ThistleSifter** [comment]
One thing that I'm not sure about with the trusted code thing is what happens when a user (let's say someone with a significant number of scripts) upgrades to the next version of RGPLua and finds that a portion or all of their scripts don't work. Unless there's something I've missed, they'd be left with two options: find and redownload all affected scripts from the website or wherever they got them from (which would be a pain), or check the trusted code setting and make the problem go away (which somewhat defeats the purpose of having the setting and trying to keep it for people who know what they're doing).

**Comment by rpatters1** [comment]
They'll have to update their scripts to the latest. I don't see a way around it.

**Comment by rpatters1** [comment]
One way to mitigate it is to ratchet up the security in stages. So we implement the trusted code mechanism first, but then we don't actually enforce it until a later version of RGP Lua.

**Comment by rpatters1** [comment]
I've been thinking about this further. "Trustedness" is going to affect more than access to metatables. It's also going to affect access to `os.execute`, and many of the `luaosutils` functions.

We will maintain a whitelist of trusted urls somewhere in the finale-lua org on Github where RGP Lua can download it easily. Then any script that wants to be trusted adds `finaleplugin.Id` to its `plugindef` function that is the url where its hash can be downloaded. RGP Lua will make sure the URL starts with a trusted URL pattern and use that to confirm the hash.

As for transitioning current users, the next version of RGP Lua will have an "opt-in" to trustedness checking. The next version after that (some months later) will start requiring it. Hopefully that gives time for users to transition to versions of code and scripts that work with trustedness before they are required to have it. I will include notices on Facebook and the jwlua list as well advising people to update everything.

**Comment by ThistleSifter** [comment]
Currently the `mixin` library checks the `personal_mixin` directory for the existence of a mixin before defaulting to the `mixin` directory.

How will trustedness affect `require`d modules from external .lua files (eg `personal_mixin`)? Will that make a script untrusted or does something else happen?

**Comment by rpatters1** [comment]
I don't think we should allow that in the `dist` directory. This whole scheme depends on trusted scripts being self-contained. If you want to run with a personal_mixin you should probably be setting the `Trusted` flag. If you are running your code out of `src` presumably you know what you are doing. The `dist` directory is for non-technical users.

**Comment by ThistleSifter** [comment]
Okay, I'll try and think of a way of detecting if the script is bundled without resorting to modifying the bundler. Maybe just a check for the existence of `package.preload["library.mixin"]` .

The only other thing I can think of is that mixins are still very much WIP, so they get updated fairly regularly which triggers a fresh bundling every time. It would be good if that didn't affect the trusted status of a user's existing local copy of a script.

**Comment by rpatters1** [comment]
Yes, I'm planning on tracking if a hash is verified and not going back to the website if it is. But we should at least if we should keep older versions of hashes. There are advantages and disadvantages to that approach.

**Comment by rpatters1** [comment]
Now that we are moving towards a concept of trusted code in RGP Lua, I think it is safe to close this issue. The trusted code conversation continues at #507.
